### ISSUE TITLE
Can not "download artifact" upgrading the agent from version [2.174.3] to version [2.188.3] 

### ISSUE THREAD

**Issue by emanuelecappelloni** [microsoft/azure-pipelines-agent#3451]
## Agent Version and Platform
Version of your agent? 2.188.3

OS of the machine running the agent? Windows

## Azure DevOps Type and Version
dev.azure.com (https://corteconti.visualstudio.com) 

## What's not working?
When upgrading the agent from version [2.174.3] to version 2.188.3 we get the error System.IO.DirectoryNotFoundException: Directory not found: 'C: \agent\externals\git\mingw64\bin' during "download artifacts task"

Inside build agent path is C: \agent\externals\git\mingw32\bin'

## Agent and Worker's Diagnostic Logs
2021-06-25T18:30:27.9558682Z ##[section]Starting: Download Artifacts
2021-06-25T18:30:27.9993454Z Creating artifacts directory: C:\agent\_work\r44\a
2021-06-25T18:30:28.0099416Z Created artifacts directory: C:\agent\_work\r44\a
2021-06-25T18:30:28.0099717Z Downloading artifact
2021-06-25T18:30:28.0115725Z Starting artifacts download...
2021-06-25T18:30:28.0127311Z Downloading linked artifact CommonScripts of type TFGit...
2021-06-25T18:30:28.0147946Z Ensuring artifact folder C:\agent\_work\r44\a\CommonScripts exists and is clean.
2021-06-25T18:30:28.0283836Z Syncing repository: 110930c0-c6f6-4a9b-9edd-43093ff11a34 (Git)
2021-06-25T18:30:28.0322853Z Prepending Path environment variable with directory containing 'git.exe'.
2021-06-25T18:30:28.0550179Z An error occurred during download: System.IO.DirectoryNotFoundException: Directory not found: 'C:\agent\externals\git\mingw64\bin'
   at Microsoft.VisualStudio.Services.Agent.Util.ArgUtil.Directory(String directory, String name)
   at Microsoft.VisualStudio.Services.Agent.Util.PathUtil.PrependPath(String directory)
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.GitCommandManager.LoadGitExecutionInfo(IExecutionContext context, Boolean useBuiltInGit)
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.GitSourceProvider.GetSourceAsync(IExecutionContext executionContext, ServiceEndpoint endpoint, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.Artifacts.TfsGitArtifact.DownloadAsync(IExecutionContext executionContext, ArtifactDefinition artifactDefinition, String downloadFolderPath)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.ReleaseJobExtension.<>c__DisplayClass39_2.<<DownloadArtifacts>b__2>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.RetryExecutor.ExecuteAsync(Func`1 action)
2021-06-25T18:30:28.0552226Z Retrying download...
2021-06-25T18:30:28.0573119Z Ensuring artifact folder C:\agent\_work\r44\a\CommonScripts exists and is clean.
2021-06-25T18:30:28.0586974Z Syncing repository: 110930c0-c6f6-4a9b-9edd-43093ff11a34 (Git)
2021-06-25T18:30:28.0594428Z Prepending Path environment variable with directory containing 'git.exe'.
2021-06-25T18:30:28.0601234Z An error occurred during download: System.IO.DirectoryNotFoundException: Directory not found: 'C:\agent\externals\git\mingw64\bin'
   at Microsoft.VisualStudio.Services.Agent.Util.ArgUtil.Directory(String directory, String name)
   at Microsoft.VisualStudio.Services.Agent.Util.PathUtil.PrependPath(String directory)
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.GitCommandManager.LoadGitExecutionInfo(IExecutionContext context, Boolean useBuiltInGit)
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.GitSourceProvider.GetSourceAsync(IExecutionContext executionContext, ServiceEndpoint endpoint, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.Artifacts.TfsGitArtifact.DownloadAsync(IExecutionContext executionContext, ArtifactDefinition artifactDefinition, String downloadFolderPath)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.ReleaseJobExtension.<>c__DisplayClass39_2.<<DownloadArtifacts>b__2>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.RetryExecutor.ExecuteAsync(Func`1 action)
2021-06-25T18:30:28.0603421Z Retrying download...
2021-06-25T18:30:28.0607658Z Ensuring artifact folder C:\agent\_work\r44\a\CommonScripts exists and is clean.
2021-06-25T18:30:28.0616954Z Syncing repository: 110930c0-c6f6-4a9b-9edd-43093ff11a34 (Git)
2021-06-25T18:30:28.0623357Z Prepending Path environment variable with directory containing 'git.exe'.
2021-06-25T18:30:28.0629679Z An error occurred during download: System.IO.DirectoryNotFoundException: Directory not found: 'C:\agent\externals\git\mingw64\bin'
   at Microsoft.VisualStudio.Services.Agent.Util.ArgUtil.Directory(String directory, String name)
   at Microsoft.VisualStudio.Services.Agent.Util.PathUtil.PrependPath(String directory)
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.GitCommandManager.LoadGitExecutionInfo(IExecutionContext context, Boolean useBuiltInGit)
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.GitSourceProvider.GetSourceAsync(IExecutionContext executionContext, ServiceEndpoint endpoint, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.Artifacts.TfsGitArtifact.DownloadAsync(IExecutionContext executionContext, ArtifactDefinition artifactDefinition, String downloadFolderPath)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.ReleaseJobExtension.<>c__DisplayClass39_2.<<DownloadArtifacts>b__2>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.RetryExecutor.ExecuteAsync(Func`1 action)
2021-06-25T18:30:28.0632890Z Retrying download...
2021-06-25T18:30:28.0637204Z Ensuring artifact folder C:\agent\_work\r44\a\CommonScripts exists and is clean.
2021-06-25T18:30:28.0645566Z Syncing repository: 110930c0-c6f6-4a9b-9edd-43093ff11a34 (Git)
2021-06-25T18:30:28.0651943Z Prepending Path environment variable with directory containing 'git.exe'.
2021-06-25T18:30:28.0659063Z An error occurred during download: System.IO.DirectoryNotFoundException: Directory not found: 'C:\agent\externals\git\mingw64\bin'
   at Microsoft.VisualStudio.Services.Agent.Util.ArgUtil.Directory(String directory, String name)
   at Microsoft.VisualStudio.Services.Agent.Util.PathUtil.PrependPath(String directory)
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.GitCommandManager.LoadGitExecutionInfo(IExecutionContext context, Boolean useBuiltInGit)
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.GitSourceProvider.GetSourceAsync(IExecutionContext executionContext, ServiceEndpoint endpoint, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.Artifacts.TfsGitArtifact.DownloadAsync(IExecutionContext executionContext, ArtifactDefinition artifactDefinition, String downloadFolderPath)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.ReleaseJobExtension.<>c__DisplayClass39_2.<<DownloadArtifacts>b__2>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.RetryExecutor.ExecuteAsync(Func`1 action)
2021-06-25T18:30:28.0661564Z Retrying download...
2021-06-25T18:30:28.0665680Z Ensuring artifact folder C:\agent\_work\r44\a\CommonScripts exists and is clean.
2021-06-25T18:30:28.0701267Z Syncing repository: 110930c0-c6f6-4a9b-9edd-43093ff11a34 (Git)
2021-06-25T18:30:28.0708091Z Prepending Path environment variable with directory containing 'git.exe'.
2021-06-25T18:30:28.0747597Z ##[error]Downloading artifacts failed: System.IO.DirectoryNotFoundException: Directory not found: 'C:\agent\externals\git\mingw64\bin'
   at Microsoft.VisualStudio.Services.Agent.Util.ArgUtil.Directory(String directory, String name)
   at Microsoft.VisualStudio.Services.Agent.Util.PathUtil.PrependPath(String directory)
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.GitCommandManager.LoadGitExecutionInfo(IExecutionContext context, Boolean useBuiltInGit)
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.GitSourceProvider.GetSourceAsync(IExecutionContext executionContext, ServiceEndpoint endpoint, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.Artifacts.TfsGitArtifact.DownloadAsync(IExecutionContext executionContext, ArtifactDefinition artifactDefinition, String downloadFolderPath)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.ReleaseJobExtension.<>c__DisplayClass39_2.<<DownloadArtifacts>b__2>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.RetryExecutor.ExecuteAsync(Func`1 action)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.ReleaseJobExtension.DownloadArtifacts(IExecutionContext executionContext, IList`1 agentArtifactDefinitions, String artifactsWorkingFolder)
   at Microsoft.VisualStudio.Services.Agent.Worker.Release.ReleaseJobExtension.DownloadArtifactsAndCommitsAsync(IExecutionContext executionContext, Object data)
2021-06-25T18:30:28.0801369Z ##[error]Directory not found: 'C:\agent\externals\git\mingw64\bin'
2021-06-25T18:30:28.0808685Z ##[section]Finishing: Download Artifacts

**Comment by mkosieradzki** [comment]
Definitely not the only issue with this buggy new version 2.188.3:

```
2021-06-28T20:58:34.0139030Z ##[error]The SSL connection could not be established, see inner exception.
2021-06-28T20:58:34.0145808Z ##[debug]Processed: ##vso[task.logissue type=error;]The SSL connection could not be established, see inner exception.
2021-06-28T20:58:34.0152942Z ##[debug]Processed: ##vso[task.complete result=Failed;]
2021-06-28T20:58:34.0157683Z ##[debug]   at Microsoft.VisualStudio.Services.Common.VssHttpRetryMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync(HttpRequestMessage message, HttpCompletionOption completionOption, Object userState, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync[T](HttpRequestMessage message, Object userState, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.Location.Client.LocationHttpClient.GetConnectionDataAsync(ConnectOptions connectOptions, Int64 lastChangeId, CancellationToken cancellationToken, Object userState)
   at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.GetConnectionDataAsync(ConnectOptions connectOptions, Int32 lastChangeId, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.ConnectAsync(ConnectOptions connectOptions, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.EnsureConnectedAsync(ConnectOptions optionsNeeded, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.CheckForServerUpdatesAsync(CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.FindServiceDefinitionsAsync(String serviceType, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.GetResourceLocationsAsync(CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientInstanceAsync(Type managedType, Guid serviceIdentifier, CancellationToken cancellationToken, VssHttpRequestSettings settings, DelegatingHandler[] handlers)
   at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientServiceImplAsync(Type requestedType, Guid serviceIdentifier, Func`4 getInstanceAsync, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientAsync[T](CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.TaskExtensions.SyncResult[T](Task`1 task)
   at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClient[T]()
   at Agent.Plugins.BuildServer..ctor(VssConnection connection)
   at Agent.Plugins.PipelineArtifactServer.DownloadAsyncV2(AgentTaskPluginExecutionContext context, ArtifactDownloadParameters downloadParameters, DownloadOptions downloadOptions, CancellationToken cancellationToken)
   at Agent.Plugins.BuildArtifacts.DownloadBuildArtifactTaskV1_0_0.ProcessCommandInternalAsync(AgentTaskPluginExecutionContext context, CancellationToken token)
   at Agent.PluginHost.Program.Main(String[] args)
```

If you want a better error message ;-) fix your error handling to include the inner exception.

**Comment by DaniilShmelev** [comment]
Hey @emanuelecappelloni! We're currently investigating this issue. It seems the problem is related to the fact that you use x86 build on x64 system. As a workaround, you can set `System.PreferGitFromPath` variable value to `true`

**Comment by JaydeepUniverse** [comment]
@DaniilShmelev  I have installed git on windows self-hosted agent, configured `System.PreferGitFromPath` to `true` in pipeline variable but the same issue still persists...

**Comment by craigeaw** [comment]
> Definitely not the only issue with this buggy new version 2.188.3:
> 
> ```
> 2021-06-28T20:58:34.0139030Z ##[error]The SSL connection could not be established, see inner exception.
> 2021-06-28T20:58:34.0145808Z ##[debug]Processed: ##vso[task.logissue type=error;]The SSL connection could not be established, see inner exception.
> 2021-06-28T20:58:34.0152942Z ##[debug]Processed: ##vso[task.complete result=Failed;]
> 2021-06-28T20:58:34.0157683Z ##[debug]   at Microsoft.VisualStudio.Services.Common.VssHttpRetryMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
>    at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
>    at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync(HttpRequestMessage message, HttpCompletionOption completionOption, Object userState, CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync[T](HttpRequestMessage message, Object userState, CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.Services.Location.Client.LocationHttpClient.GetConnectionDataAsync(ConnectOptions connectOptions, Int64 lastChangeId, CancellationToken cancellationToken, Object userState)
>    at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.GetConnectionDataAsync(ConnectOptions connectOptions, Int32 lastChangeId, CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.ConnectAsync(ConnectOptions connectOptions, CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.EnsureConnectedAsync(ConnectOptions optionsNeeded, CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.CheckForServerUpdatesAsync(CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.FindServiceDefinitionsAsync(String serviceType, CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.GetResourceLocationsAsync(CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientInstanceAsync(Type managedType, Guid serviceIdentifier, CancellationToken cancellationToken, VssHttpRequestSettings settings, DelegatingHandler[] handlers)
>    at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientServiceImplAsync(Type requestedType, Guid serviceIdentifier, Func`4 getInstanceAsync, CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClientAsync[T](CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.Services.WebApi.TaskExtensions.SyncResult[T](Task`1 task)
>    at Microsoft.VisualStudio.Services.WebApi.VssConnection.GetClient[T]()
>    at Agent.Plugins.BuildServer..ctor(VssConnection connection)
>    at Agent.Plugins.PipelineArtifactServer.DownloadAsyncV2(AgentTaskPluginExecutionContext context, ArtifactDownloadParameters downloadParameters, DownloadOptions downloadOptions, CancellationToken cancellationToken)
>    at Agent.Plugins.BuildArtifacts.DownloadBuildArtifactTaskV1_0_0.ProcessCommandInternalAsync(AgentTaskPluginExecutionContext context, CancellationToken token)
>    at Agent.PluginHost.Program.Main(String[] args)
> ```
> 
> If you want a better error message ;-) fix your error handling to include the inner exception.
> 
> This one is on RHEL7.

We're having the exact same issue on Windows Server 2019.

**Comment by cmooring** [comment]
> ## Agent Version and Platform
> Version of your agent? 2.188.3
> 
> OS of the machine running the agent? Windows
> 
> ## Azure DevOps Type and Version
> dev.azure.com (https://corteconti.visualstudio.com)
> 
> ## What's not working?
> When upgrading the agent from version [2.174.3] to version 2.188.3 we get the error System.IO.DirectoryNotFoundException: Directory not found: 'C: \agent\externals\git\mingw64\bin' during "download artifacts task"

We're facing the same issue after updating to 2.188.3 too.

**Comment by dbhadri** [comment]
### Agent Version and Platform
Agent Version - 2.188.3

Agent OS - Windows X64

### What's not working?

The agent can download the artifacts from the build job for a release - but it cannot download the code from the repository and fails with the following error.

Downloading artifacts failed: System.IO.DirectoryNotFoundException: Directory not found: 'C:\agent\externals\git\mingw64\bin'

We noticed that this happened after the Agents automatically updated itself from 2.186.1 to 2.188.3

**Comment by emanuelecappelloni** [comment]
Hi @DaniilShmelev.  "System.PreferGitFromPath to true" work around seems to work for me. 
Anyway  I wait a new version of the agent that solve the issue becouse i should change a lot of pipeline for fix.
Really thanks

**Comment by craigeaw** [comment]
I submitted a new issue specific to the SSL error: #3452

**Comment by seddikmahmah** [comment]
Hi,
we are facing a similar issue with this new version.
The download artifact task fails. But the download Git works correctly.
Our self-hosted agents are installed in a **RedHat7** VMs and also in an AKS (**docker ubuntu18** image) with the same problem.
The only message shown on the release is a red error text, Forbidden.
On the Woker log files (VM and Docker image) we found the next error:
"Access to the path '/proc/xxx/oom_score_adj' is denied"
The VM is a 4Gb RAM and 2vCore used only to host the agent. And when we increased the RAM to 16Gb and 4vCore, apparently the error isn't triggered on the log files, but the releases fail with the same error message.
![Captura de pantalla 2021-06-30 a las 11 16 02](https://user-images.githubusercontent.com/80770040/124045068-2d0b1e00-da0f-11eb-893d-09c0822e7215.png)
![Captura de pantalla 2021-06-30 a las 4 32 09](https://user-images.githubusercontent.com/80770040/124045085-35635900-da0f-11eb-9a74-92ff353488d1.png)

**Comment by cmooring** [comment]
> > ## Agent Version and Platform
> > Version of your agent? 2.188.3
> > OS of the machine running the agent? Windows
> > ## Azure DevOps Type and Version
> > dev.azure.com (https://corteconti.visualstudio.com)
> > ## What's not working?
> > When upgrading the agent from version [2.174.3] to version 2.188.3 we get the error System.IO.DirectoryNotFoundException: Directory not found: 'C: \agent\externals\git\mingw64\bin' during "download artifacts task"
> 
> We're facing the same issue after updating to 2.188.3 too.

We were able to work around the issue by manually downloading the 64 bit version of the agent again, and re-installing it manually. I didn't do the install myself, our operations team did it, but from what I've been told, the config.bat (ie first time install script) has an option to replace.

So for now, we're good to go, but it'd be great to have this addressed in a future release.

**Comment by tarangpandya** [comment]
Hello - We are also getting the errors while agent trying to download the artifacts. However, In our scenario my agent is behind firewall and we did whitlisted routes and IP addresses as it is described here - https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/v2-windows?view=azure-devops#for-organizations-using-the-visualstudiocom-domain

Our CI/CD process was working fine before upgrading it to 188.3 

Current Agent Version : 2.188.3
Exception : Can not download the artifacts , while agent is behind firewall.

![image](https://user-images.githubusercontent.com/23608276/124173403-fc5fce80-da78-11eb-9575-f2a0904a0d94.png)

We are getting bad gateway response code when it tries to download the artifacts.

**Comment by DaniilShmelev** [comment]
@tarangpandya hi, would you mind creating a separate ticket for your issue?

**Comment by DaniilShmelev** [comment]
@emanuelecappelloni hi, we've fixed this issue in 2.188.4. It should be rolled out in the next few days.
For now you can download the agent here: https://github.com/microsoft/azure-pipelines-agent/releases/tag/v2.188.4
